### PR TITLE
Add AVX2 instruction assertiosn

### PR DESCRIPTION
**Pull request by alexcrichton**
Also a few other assorted modules

### CODE REVIEW

File: src/x86/avx2.rs [hunk 1]
```diff
@@ -621,6 +687,7 @@ pub fn _mm256_sll_epi64(a: i64x4, count: i64x2) -> i64x4 {
 /// shifting in zeros, return the results;
 #[inline(always)]
 #[target_feature = "+avx2"]
+#[cfg_attr(test, assert_instr(vpsllw))] // TODO: should this be pslli
```
Comment (alexcrichton) [inline src/x86/avx2.rs]: @BurntSushi do you know what's up with this? (and the two below)

File: src/x86/avx2.rs [hunk 2]
```diff
@@ -699,6 +774,7 @@ pub fn _mm256_sra_epi32(a: i32x8, count: i32x4) -> i32x8 {
 /// shifting in sign bits.
 #[inline(always)]
 #[target_feature = "+avx2"]
+#[cfg_attr(test, assert_instr(vpsraw))] // TODO: notvpsraiw?
```
Comment (alexcrichton) [inline src/x86/avx2.rs]: Same with this, and the one below

File: src/x86/avx2.rs [hunk 3]
```diff
@@ -756,6 +838,7 @@ pub fn _mm256_srl_epi64(a: i64x4, count: i64x2) -> i64x4 {
 /// zeros
 #[inline(always)]
 #[target_feature = "+avx2"]
+#[cfg_attr(test, assert_instr(vpsrlw))] // TODO not vpsrliw?
```
Comment (alexcrichton) [inline src/x86/avx2.rs]: And this

**Comment by alexcrichton** [comment]
Sounds like the minor inconsistencies here have a reason but we've forgotten why. Sounds like a comment to add at a future date!

File: src/x86/avx2.rs (hunk 1 above)
Comment (BurntSushi) [reply src/x86/avx2.rs]: It looks like Clang's codegen tests think that `pslli` should be emitted, at least in the llvm IR: https://github.com/llvm-mirror/clang/blob/6d55fdd9ee95cdb0e6fd34df57e044faff8fe5f0/test/CodeGen/avx2-builtins.c#L988-L989

Comment (BurntSushi) [reply src/x86/avx2.rs]: But it looks like the llvm codegen tests agree with what we're observing, and specifically that pslli in the ir gets lowered to vpsllw: https://github.com/llvm-mirror/llvm/blob/4c92030df73fdfac8b57d86241d26796f7f35a80/test/CodeGen/X86/avx2-intrinsics-fast-isel.ll#L2707-L2754

**Pull request merged**